### PR TITLE
YAML consolidation: `netplan set` (FR-702)

### PR DESCRIPTION
**Pull request by schopin-pro**
## Description

As part of the larger PyYAML cleanup effort, this patch set rewrites the `netplan set` to use libnetplan's own YAML parser instead of PyYAML. This meant enriching the parser to deal with null fields, that need to be tracked out-of-band, as well as creating new APIs to create YAML files out of a Netplan state, with different nuances to match the previous behavior.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).

### CODE REVIEW

**Comment by schopin-pro** [comment]
This builds up on #252 hence the draft status.

File: src/netplan.c [hunk 1]
```diff
@@ -991,6 +991,52 @@ netplan_netdef_list_write_yaml(const NetplanState* np_state, GList* netdefs, int
...
+    if (to_write == NULL)
+        return TRUE;
+
+    int out_fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, 0640);
+    if (out_fd < 0) {
+        g_set_error(error, G_FILE_ERROR, errno, "%s", strerror(errno));
+        return FALSE;
+    }
+
+    gboolean ret = netplan_netdef_list_write_yaml(np_state, to_write, out_fd, error);
```
Comment (schopin-pro) [inline src/netplan.c:1035]: `to_write` leaks.

**Comment by schopin-pro** [comment]
Ran the autopkgtests manually:

```
autopkgtest [17:20:13]: @@@@@@@@@@@@@@@@@@@@ summary
ovs                  PASS
ethernets            PASS
bridges              PASS
bonds                PASS
routing              PASS
vlans                PASS
wifi                 PASS
tunnels              PASS
scenarios            PASS
regressions          PASS
autostart            PASS
cloud-init           PASS
```

File: src/util.c [hunk 2]
```diff
@@ -523,10 +523,10 @@ netplan_delete_connection(const char* id, const char* rootdir)
     del = g_strdup_printf("network.%s.%s=NULL", netplan_def_type_name(nd->type), id);
 
     /* TODO: refactor logic to actually be inside the library instead of spawning another process */
-    const gchar *argv[] = { SBINDIR "/" "netplan", "set", del, "--origin-hint" , filename, NULL, NULL, NULL };
```
Comment (slyon) [inline src/util.c:418]: We should delete the `filename` variable processing above, as we don't make any use of it anymore.
Especially this part:
```
    filename = g_path_get_basename(nd->filename);
    filename[strlen(filename) - 5] = '\0'; //stip ".yaml" suffix
```

File: tests/test_cli_get_set.py [hunk 3]
```diff
@@ -63,53 +63,54 @@ def test_set_scalar(self):
...
+            self.assertIs(True, yaml.safe_load(f)['network']['ethernets']['eth0']['dhcp4'])
 
     def test_set_scalar2(self):
         self._set(['ethernets.eth0.dhcp4="yes"'])
         self.assertTrue(os.path.isfile(self.path))
         with open(self.path, 'r') as f:
-            self.assertIn('network:\n  ethernets:\n    eth0:\n      dhcp4: \'yes\'', f.read())
+            # XXX: the previous version using PyYAML would keep the "yes" variant but the round-trip
+            # through libnetplan doesn't keep formatting choices (yes is a keyword same as true)
+            self.assertIs(True, yaml.safe_load(f)['network']['ethernets']['eth0']['dhcp4'])
```
Comment (slyon) [inline tests/test_cli_get_set.py:74]: thought (non-blocking): Could we somehow make use of the new "dirty" tracking functionality to keep formatting choices?

File: tests/test_cli_get_set.py [hunk 4]
```diff
@@ -192,15 +193,17 @@ def test_set_append(self):
             f.write('''network:
   version: 2
   ethernets:
+    ens2: {dhcp4: yes}
```
Comment (slyon) [inline tests/test_cli_get_set.py:196]: nitpick (non-blocking): This looks unrelated and unused, why was it introduced?

File: tests/test_cli_get_set.py [hunk 5]
```diff
@@ -210,39 +213,42 @@ def test_set_overwrite_eq(self):
...
+    ens3: {dhcp4: "no"}''')
         self._set(['ethernets.ens3.dhcp4=true'])
         self.assertTrue(os.path.isfile(self.path))
         with open(self.path, 'r') as f:
-            out = f.read()
-            self.assertIn('network:\n  ethernets:\n', out)
-            self.assertIn('    ens3:\n      dhcp4: true', out)
+            self.assertIs(
+                    True,
+                    yaml.safe_load(f)['network']['ethernets']['ens3']['dhcp4'])
```
Comment (slyon) [inline tests/test_cli_get_set.py:230]: (non-blocking): We could change the formatting to make this look a bit nicer:
```
out = yaml.safe_load(f)
self.assertTrue(out['network']['ethernets']['ens3']['dhcp4'])
```

File: tests/test_cli_get_set.py [hunk 6]
```diff
@@ -300,9 +432,9 @@ def test_set_override_existing_file(self):
         self.assertFalse(os.path.isfile(self.path))
         self.assertTrue(os.path.isfile(override))
         with open(override, 'r') as f:
-            out = f.read()
-            self.assertIn('network:\n  ethernets:\n    eth0:\n      dhcp4: false', out)  # new
-            self.assertIn('eth1:\n      dhcp6: false', out)  # old
+            out = yaml.safe_load(f)
+            self.assertIs(False, out['network']['ethernets']['eth0']['dhcp4'])
+            self.assertIs(False, out['network']['ethernets']['eth1']['dhcp6'])
```
Comment (slyon) [inline tests/test_cli_get_set.py:437]: (non-blocking): I think we could be using `assertFalse()` here and below.

File: tests/test_cli_get_set.py [hunk 7]
```diff
@@ -321,25 +454,25 @@ def test_set_override_multiple_existing_files(self):
...
-                    r'eth0.2:{dhcp4: false}},'
-                    r'bridges:{'
-                    r'br99:{dhcp4: false}}}')])
+        self._set([(r'network={renderer: NetworkManager, version: 2, '
+                    r'ethernets: {'
+                    r'eth1: {dhcp4: false},'
+                    r'eth0.1: {dhcp4: false},'
+                    r'eth0.2: {dhcp4: false}},'
+                    r'bridges: {'
+                    r'br99: {dhcp4: false}}}')])
```
Comment (slyon) [inline tests/test_cli_get_set.py:463]: question: Is this just some cleanup or an actual change in behavior, that we now require properly space separated mappings? We would need to take care not to break any consumer (like our own `netplan-dbus` daemon and/or `snapd`'s get/set functionality).

**Review by slyon — COMMENTED** [review]
This is another big PR, but the change makes the implementation so much cleaner and more explicit. I really like your overall approach!

I left several comments inline about some details that we should address in a 2nd version of PR.

**Comment by schopin-pro** [comment]
For the sake of being able to test stuff, I included the commit of #257 since without it the null-handling breaks stuff.

File: src/parse.c [hunk 8]
```diff
@@ -2907,3 +2986,61 @@ netplan_parser_clear(NetplanParser** npp_p)
...
+    yaml_node_pair_t* entry;
+    switch (node->type) {
+        // LCOV_EXCL_START
+        case YAML_NO_NODE:
+            g_hash_table_insert(null_fields, key_prefix, key_prefix);
+            key_prefix = NULL;
+            break;
+        // LCOV_EXCL_STOP
+        case YAML_SCALAR_NODE:
+            if (       g_ascii_strcasecmp("null", scalar(node)) == 0
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/parse.c:3041]: We already do. `g_ascii_strcasecmp` is case-insensitive.

File: tests/test_cli_get_set.py (hunk 6 above)
Comment (schopin-pro) [reply tests/test_cli_get_set.py:437]: Sadly, no. `assertFalse` and `assertTrue` coerce the tested value into a boolean, and we want to make sure we have the actual boolean values here, not any truthy or falsy value.

File: tests/test_cli_get_set.py [hunk 9]
```diff
@@ -63,53 +63,54 @@ def test_set_scalar(self):
         self._set(['ethernets.eth0.dhcp4=true'])
         self.assertTrue(os.path.isfile(self.path))
         with open(self.path, 'r') as f:
-            self.assertIn('network:\n  ethernets:\n    eth0:\n      dhcp4: true', f.read())
+            self.assertIs(True, yaml.safe_load(f)['network']['ethernets']['eth0']['dhcp4'])
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply tests/test_cli_get_set.py:66]: As commented elsewhere, there's a subtle difference between `assertIs(True, ` and `assertTrue(`, the latter doing a boolean conversion, which we don't want here.

File: tests/test_cli_get_set.py (hunk 3 above)
Comment (schopin-pro) [reply tests/test_cli_get_set.py:74]: I suppose we could, by storing the actual scalar data as the value in the hashtable. I'm not sure it's worth the hassle, though.

File: tests/test_cli_get_set.py (hunk 4 above)
Comment (schopin-pro) [reply tests/test_cli_get_set.py:196]: Probably a leftover of some debugging session or another. I'll drop it.

File: tests/test_cli_get_set.py (hunk 7 above)
Comment (schopin-pro) [reply tests/test_cli_get_set.py:463]: Oh crap. It seems it is a change in behavior, I get `netplan.libnetplan.LibNetplanException: Error parsing YAML: found unexpected ':'` when reverting the changes.
This seems weird to me, as the underlying parsing library is the same, isn't it?

FWIW, I checked the YAML spec, and at least in 1.1 and 1.2 the whitespace after the colon is mandatory.

Comment (slyon) [reply tests/test_cli_get_set.py:463]: I don't know for sure if PyYAML is using libyaml in the background, too. If those two YAML implementations differ that would be a problem, especially as we have consumers using that (out of spec?) syntax: https://github.com/snapcore/snapd/pull/10752/files#diff-e380767e9d06e93d76ea779341439e7d4c878b2708bcc23731453c17c2442616R557

I wonder if we can somehow handle both cases and maybe print a warning that this does not comply with the YAML specification?

Comment (schopin-pro) [reply tests/test_cli_get_set.py:463]: After investigations, I note the following things:

* Even though PyYAML loads libyaml, it does *not* use it in the default case. You have to explicitly use the CLoader class in order to do so. By default it uses a pure Python parser, which is apparently less strict than the C one.
* When using CLoader on the original string, I get the same error.
* However, when using the one in the snapd source code, it loads just fine. The difference is that the snapd code quotes their keys, since they're using a JSON serializer to format their string (which is valid since JSON is a subset of YAML). So, on that front I think we're OK.

File: src/parse.c [hunk 10]
```diff
@@ -187,7 +233,7 @@ netplan_netdef_new(NetplanParser *npp, const char* id, NetplanDefType type, Netp
 
 typedef gboolean (*node_handler) (NetplanParser* npp, yaml_node_t* node, const void* data, GError** error);
 
-typedef gboolean (*custom_map_handler) (NetplanParser* npp, yaml_node_t* node, const void* data, GError** error);
+typedef gboolean (*custom_map_handler) (NetplanParser* npp, yaml_node_t* node, const char *prefix, const void* data, GError** error);
```
Comment (slyon) [inline src/parse.c:236]: key_prefix?

File: src/parse.c [hunk 11]
```diff
@@ -2873,6 +3021,17 @@ netplan_parser_reset(NetplanParser* npp)
     }
 
     npp->missing_ids_found = 0;
+
+    if (npp->null_fields) {
+        g_hash_table_destroy(npp->null_fields);
+        npp->null_fields = NULL;
+    }
+
+    if(npp->sources) {
```
Comment (slyon) [inline src/parse.c:3030]:
```suggestion
    if (npp->sources) {
```

File: src/types.c [hunk 12]
```diff
@@ -384,6 +394,12 @@ netplan_state_reset(NetplanState* np_state)
 
     np_state->backend = NETPLAN_BACKEND_NONE;
     reset_ovs_settings(&np_state->ovs_settings);
+
+    if(np_state->sources) {
```
Comment (slyon) [inline src/types.c:398]:
```suggestion
    if (np_state->sources) {
```

File: tests/test_cli_get_set.py (hunk 3 above)
Comment (slyon) [reply tests/test_cli_get_set.py:59]: ACK. we could still do this in the future, if needed. Let's keep it out of this (already big) PR.

**Review by slyon — COMMENTED** [review]
Thank you very much, we're making good progress with this big PR and are getting much closer to a unified YAML parser. kudos!

I've left some smaller inline comments, nothing critical.

Another thing to mention is that `make check-coverage` currently fails, due to `netplan/cli/sriov.py:363`, which can probably just be marked as "nocover", as it's just about (non-existing) hardware specific quirks.

File: tests/test_cli_get_set.py (hunk 7 above)
Comment (schopin-pro) [reply tests/test_cli_get_set.py]: To close up this thread: I've added a new commit that explicitly switches this to a JSON-payload to match the snapd use-case.

File: src/netplan.c [hunk 13]
```diff
@@ -989,6 +989,53 @@ netplan_netdef_list_write_yaml(const NetplanState* np_state, GList* netdefs, int
...
+gboolean
+netplan_state_write_yaml_file(const NetplanState* np_state, const char* filename, const char* rootdir, GError** error)
+{
+    GList* iter = np_state->netdefs_ordered;
+    g_autofree gchar *path = NULL;
+    GList* to_write = NULL;
+
+    path = g_build_path(G_DIR_SEPARATOR_S, rootdir ?: G_DIR_SEPARATOR_S, "etc", "netplan", filename, NULL);
+
+    if (unlink(path) && errno != ENOENT) {
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c:1011]: Thanks for this, that was a good idea! I used a slightly different approach with similar results: create a temporary file /etc/netplan/FILENAME.yaml.$RANDOM, write stuff to it, and rename it to the proper filename if all is well.

File: src/netplan.c [hunk 14]
```diff
@@ -1039,6 +1039,76 @@ netplan_state_dump_yaml(const NetplanState* np_state, int out_fd, GError** error
...
+            g_hash_table_insert(perfile_netdefs, default_path, NULL);
+        }
+    } else {
+        GList* iter = np_state->netdefs_ordered;
+        while (iter) {
+            NetplanNetDefinition* netdef = iter->data;
+            const char* filename = netdef->filename ? netdef->filename : default_path;
+            GList* list = NULL;
+            g_hash_table_steal_extended(perfile_netdefs, filename, NULL, (gpointer*)&list);
+            g_hash_table_insert(perfile_netdefs, (gpointer)filename, g_list_append(list, netdef));
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c:1087]: NAK. Except for `default_path`, our keys are owned by a `netdef` object so we simply cannot have any destructor function for keys. We *could* strdup them, but IMHO the current code works fine as is. Furthermore, we need to actually do the steal + insert, as we need the previous value to construct the new one *and* we really don't want to run the destructor on it.

File: src/netplan.c [hunk 15]
```diff
@@ -1050,3 +1050,76 @@ netplan_state_dump_yaml(const NetplanState* np_state, int out_fd, GError** error
...
+ */
+gboolean
+netplan_state_update_yaml_hierarchy(const NetplanState* np_state, const char* default_filename, const char* rootdir, GError** error)
+{
+    g_autofree gchar *default_path = NULL;
+    gboolean ret = FALSE;
+    GHashTableIter hash_iter;
+    gpointer key, value;
+    GHashTable *perfile_netdefs = g_hash_table_new_full(g_str_hash, g_str_equal, NULL, (GDestroyNotify)g_list_free);
+    default_path = g_build_path(G_DIR_SEPARATOR_S, rootdir ?: G_DIR_SEPARATOR_S, "etc", "netplan", default_filename, NULL);
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c:1071]: Fixed by using g_assert (since we do it elsewhere for netdef validation).

File: netplan/cli/commands/set.py [hunk 16]
```diff
@@ -54,151 +49,37 @@ def run(self):
...
-            if os.path.isfile(absp):
-                os.remove(absp)
-        else:  # pragma nocover
-            raise Exception('Invalid input: {}'.format(set_tree))
+        key, value = split
+        if not key.startswith('network'):
+            key = '.'.join(('network', key))
+
+        # Replace the '.' with '\t' but not at '\.' via negative lookbehind expression
+        yaml_path = [s.replace(r'\.', '.') for s in re.split(r'(?<!\\)\.', key)]
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply netplan/cli/commands/set.py]: No we're not, we're actually splitting the string into a list, and replacing the escaped dots by their standard counterpart. The '\t' encoding is done in the bindings, as it's an implementation detail of the C library.

I fixed the comment :)

File: src/netplan.c [hunk 17]
```diff
@@ -1050,3 +1050,76 @@ netplan_state_dump_yaml(const NetplanState* np_state, int out_fd, GError** error
...
+    }
+
+    /* Remove any referenced source file that doesn't have any associated data.
+       Presumably, it is data that has been obsoleted by files loaded
+       afterwards, typically via `netplan set`. */
+    if (np_state->sources) {
+        g_hash_table_iter_init(&hash_iter, np_state->sources);
+        while (g_hash_table_iter_next (&hash_iter, &key, &value)) {
+            if (!g_hash_table_contains(perfile_netdefs, key))
+                g_unlink(key);
```
[the remark this answers is not in the document]
Comment (schopin-pro) [reply src/netplan.c]: Done, with special casing for the "file does not exist" error which is ignored.

**Review by slyon — APPROVED** [review]
Thank you! All my remarks have been addressed. This is ready to be merged IMO.

Abigail is detecting some ABI changes, but they are all compatible (added functions & appended `sources` to `NetplanState`). Feel free to update the ABI XML to something like this (after a rebase, to have `abi-compat/focal_0.104.xml` available): https://paste.ubuntu.com/p/BK8HGnzPrG/ and merge this PR afterwards.
```
abidiff abi-compat/focal_0.104.xml libnetplan.so.0.0 --headers-dir2 include/ --header-file2 src/abi.h
Functions changes summary: 0 Removed, 0 Changed (26 filtered out), 5 Added functions
Variables changes summary: 0 Removed, 1 Changed, 0 Added variable

5 Added functions:

  [A] 'function gboolean netplan_parser_load_nullable_fields(NetplanParser*, int, GError**)'    {netplan_parser_load_nullable_fields}
  [A] 'function gboolean netplan_parser_load_yaml_from_fd(NetplanParser*, int, GError**)'    {netplan_parser_load_yaml_from_fd}
  [A] 'function gboolean netplan_state_update_yaml_hierarchy(const NetplanState*, const char*, const char*, GError**)'    {netplan_state_update_yaml_hierarchy}
  [A] 'function gboolean netplan_state_write_yaml_file(const NetplanState*, const char*, const char*, GError**)'    {netplan_state_write_yaml_file}
  [A] 'function gboolean netplan_util_create_yaml_patch(const char*, const char*, int, GError**)'    {netplan_util_create_yaml_patch}

1 Changed variable:

  [C] 'NetplanState global_state' was changed at abi_compat.c:6[4](https://github.com/canonical/netplan/runs/6174614861?check_suite_focus=true#step:5:4):1:
    size of symbol changed from 1[6](https://github.com/canonical/netplan/runs/6174614861?check_suite_focus=true#step:5:6)8 to 1[7](https://github.com/canonical/netplan/runs/6174614861?check_suite_focus=true#step:5:7)6
```

**Comment by slyon** [comment]
> IMHO we should completely exclude `global_state` from the ABI check

ACK. Should be resolved by a4a8ac8264c613315b314d3c395dd0985fa63743

File: src/netplan.c [hunk 18]
```diff
@@ -989,6 +989,64 @@ netplan_netdef_list_write_yaml(const NetplanState* np_state, GList* netdefs, int
...
+        if (g_strcmp0(fname, path) == 0)
+            to_write = g_list_append(to_write, netdef);
+        iter = iter->next;
+    }
+
+    /* Remove any existing file if there is no data to write */
+    if (to_write == NULL) {
+        if (unlink(path) && errno != ENOENT) {
+            g_set_error(error, G_FILE_ERROR, errno, "%s", strerror(errno));
+            return FALSE;
```
Comment (slyon) [inline src/netplan.c:1025]: This is not covered with the current testing coverage. Otherwise, I think we're good to merge this after a4a8ac8264c613315b314d3c395dd0985fa63743 landed

File: src/netplan.c [hunk 19]
```diff
@@ -1003,3 +1061,83 @@ netplan_state_dump_yaml(const NetplanState* np_state, int out_fd, GError** error
...
+
+    /* Remove any referenced source file that doesn't have any associated data.
+       Presumably, it is data that has been obsoleted by files loaded
+       afterwards, typically via `netplan set`. */
+    if (np_state->sources) {
+        g_hash_table_iter_init(&hash_iter, np_state->sources);
+        while (g_hash_table_iter_next (&hash_iter, &key, &value)) {
+            if (!g_hash_table_contains(perfile_netdefs, key)) {
+                if (unlink(key) && errno != ENOENT)
+                    goto file_error;
```
Comment (slyon) [inline src/netplan.c:1128]: Same here, needs coverage or exclusion.

Comment (schopin-pro) [reply src/netplan.c]: I went for exclusion for this one as testing it out was really convoluted for relatively little gain.

**Comment by schopin-pro** [comment]
Regarding editing the ABI doc, shouldn't we hold until release? That way we can still tweak symbols introduced in previous PRs

**Comment by slyon** [comment]
> Regarding editing the ABI doc, shouldn't we hold until release? That way we can still tweak symbols introduced in previous PRs

Yes, makes sense! Feel free to merge this PR.
